### PR TITLE
todo.txt: Better project/context management

### PR DESCRIPTION
**Pull request by harshad1**
Reworked the add context and project dialogs.

1. When the action is clicked the existing contexts / projects are shown
2. Unselecting or selecting any and clicking ok will remove the de-selected ones and add the selected ones
3. Chunking support added so this entire operation is atomic and can be undone with one press of the undo button

TBD:
- [x] Cleanups (especially imports)
- [x] Some perf tweaks
- [x] More testing to determine / fix edge cases

### CODE REVIEW

File: app/src/main/java/net/gsantner/markor/format/ActionButtonBase.java [hunk 1]
```diff
@@ -392,16 +393,20 @@ public static void runRegexReplaceAction(final EditText editor, final ReplacePat
...
-            ((HighlightingEditor) editor).withAutoFormatDisabled(() -> _runRegexReplaceAction(editor, patterns));
+            ((HighlightingEditor) editor).withAutoFormatDisabled(() -> runRegexReplaceAction(editor.getText(), patterns));
         } else {
-            _runRegexReplaceAction(editor, patterns);
+            runRegexReplaceAction(editor.getText(), patterns);
         }
     }
 
-    private static void _runRegexReplaceAction(final EditText editor, final List<ReplacePattern> patterns) {
+    public static void runRegexReplaceAction(final Editable editable, final ReplacePattern... patterns) {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/format/ActionButtonBase.java:402]: Reworked so it works on Editable, not the view. More general now.

File: app/src/main/java/net/gsantner/markor/frontend/MarkorDialogFactory.java [hunk 2]
```diff
@@ -572,6 +576,34 @@ public static void showInsertItemsDialog(
         GsSearchOrCustomTextDialog.showMultiChoiceDialogWithSearchFilterUI(activity, dopt);
     }
 
+    // Insert items
+    public static void showUpdateItemsDialog(
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/frontend/MarkorDialogFactory.java:580]: General dialog for updating items

File: app/src/main/java/net/gsantner/markor/frontend/textview/TextViewUtils.java [hunk 3]
```diff
@@ -435,28 +442,44 @@ public static class ChunkedEditable implements Editable {
...
 
         // Apply changes from copy to original
-        public boolean applyChanges() {
+        public void applyChanges() {
             if (copy == null) {
-                return false;
+                return;
+            }
+
+            if (_depth > 0) {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/frontend/textview/TextViewUtils.java:471]: Used to make stacking and unwinding simple.

File: app/src/main/java/net/gsantner/markor/frontend/textview/TextViewUtils.java [hunk 4]
```diff
@@ -567,27 +600,27 @@ public InputFilter[] getFilters() {
 
         @Override
         public void setSpan(Object what, int start, int end, int flags) {
-            // Do nothing
+            setSel(what, start);
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/frontend/textview/TextViewUtils.java:603]: Very limited span support so that we can set and get selection from a ChunkedEditable.

File: app/src/main/java/net/gsantner/markor/frontend/textview/TextViewUtils.java [hunk 5]
```diff
@@ -667,4 +700,47 @@ public static Boolean isImeOpen(final View view) {
...
+            }
+        }
+    }
+
+
+    public interface MTextWatcher extends TextWatcher {
+        // Classes inheriting from this class will be disabled in withAutoFormatDisabled
+    }
+
+    public static void withAutoFormatDisabled(final Editable editable, final GsCallback.a0 callback) {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/markor/frontend/textview/TextViewUtils.java:722]: TBD make this passthrough for a ChunkedEditable for performance

File: app/src/main/java/net/gsantner/opoc/frontend/GsSearchOrCustomTextDialog.java [hunk 6]
```diff
@@ -384,8 +385,9 @@ public static void showMultiChoiceDialogWithSearchFilterUI(final Activity activi
             }
         });
 
-        // long click always activates
-        listView.setOnItemLongClickListener((parent, view, pos, id) -> directActivate.callback(pos));
+        if (dopt.isLongPressSelectEnabled) {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/opoc/frontend/GsSearchOrCustomTextDialog.java:388]: Option to disable long press listener. Default true

File: app/src/main/java/net/gsantner/opoc/util/GsCollectionUtils.java [hunk 7]
```diff
@@ -115,38 +120,25 @@ public static <T, K extends Comparable<K>> void keySort(
...
      */
-    public static <T> int[] diff(final List<T> dest, final List<T> source) {
-        final int dl = dest.size(), sl = source.size();
-        final int minLength = Math.min(dl, sl);
-
-        int start = 0, fromEnd = 0;
-        while (start < minLength && source.get(start).equals(dest.get(start))) {
-            start++;
-        }
+    public static <T> Set<T> sub(final Collection<T> a, final Collection<T> b) {
```
Comment (harshad1) [inline app/src/main/java/net/gsantner/opoc/util/GsCollectionUtils.java:126]: TBD - name this setDiff

**Comment by gsantner** [comment]
Updated with changes from master. As far I can tell this one is still ongoing

**Comment by harshad1** [comment]
@gsantner This is ready to merge
